tools/pre-commit

```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
+# simple pre commit hook script to check that *.cpp and *.hpp files 
+# are correctly clang-formatted and that CMakeLists.txt and *.cmake
+# files are cmake-formatted
+
+# to install this hook, copy this file to your git hooks as follows
+# cp tools/pre-commit .git/hooks/pre-commit
+# chmod +x .git/hooks/pre-commit
+
+red=$(tput setaf 1)
+green=$(tput setaf 2)
+yellow=$(tput setaf 3)
+blue=$(tput setaf 4)
+normal=$(tput sgr0)
+
+cxxfiles=()
+for file in `git diff --cached --name-only --diff-filter=ACMRT | grep -E "\.(cpp|hpp)$"`; do
+    if ! cmp -s <(git show :${file}) <(git show :${file}|clang-format -style=file); then
+        cxxfiles+=("${file}")
+    fi
+done
+
+cmakefiles=()
+for file in `git diff --cached --name-only --diff-filter=ACMRT | grep -E "(CMakeLists\.txt|\.cmake)$"`; do
+    if ! cmp -s <(git show :${file}) <(git show :${file}|cmake-format); then
+        cmakefiles+=("${file}")
+    fi
+done
+
+returncode=0
+
+if [ -n "${cxxfiles}" ]; then
+    printf "# ${blue}clang-format ${red}error pre-commit${normal} : To fix run the following (use git commit ${yellow}--no-verify${normal} to bypass)\n"
+    for f in "${cxxfiles[@]}" ; do
+        printf "clang-format -style=file -i %s\n" "$f"
```

Comment:
Maybe:
```suggestion
        printf "clang-format -i %s\n" "$f"
```
? Since it'll use the file at the repository root by default.

Reply:
`-style=file` uses the file loaded from the parent directories- I put this in to make it explicit that we will use the file from the repo root - even if the clang-format default were to change (which it won't)
```
  --style=<string>    ...
                               Use -style=file to load style configuration from
                               .clang-format file located in one of the parent
                               directories of the source file (or current
                               directory for stdin).
```

Reply:
Oh, very nice! Never mind then. Thanks!